src/DocBlock/TypeExpression.php

```diff
@@ -427,14 +446,61 @@ private function parse(): void
             if (\strlen($this->value) <= $index) {
                 \assert(\strlen($this->value) === $index);
 
-                $this->isUnionType = true;
-                $this->typesGlue = $typesGlue;
+                $this->isGluedType = true;
+
+                if (1 === \count($seenGlues)) {
+                    $this->typesGlue = array_key_first($seenGlues);
+
+                    foreach ($innerValues as $innerValue) {
+                        $this->innerTypeExpressions[] = [
+                            'start_index' => $innerValue['start_index'],
+                            'expression' => $this->inner($innerValue['value']),
+                        ];
+                    }
+                } else {
+                    $glue = null;
+                    foreach (['|', '&', '~'] as $possibleGlue) {
+                        if ($seenGlues[$possibleGlue] ?? false) {
+                            $glue = $possibleGlue;
+
+                            break;
+                        }
+                    }
+                    \assert(null !== $glue);
+
+                    $this->typesGlue = $glue;
+
+                    for ($i = 0; $i < \count($innerValues); ++$i) {
+                        $innerStartIndex = $innerValues[$i]['start_index'];
+                        $innerValue = '';
+                        while (true) {
+                            $innerValue .= $innerValues[$i]['value'];
+
+                            if (($innerValues[$i]['next_glue'] ?? $glue) === $glue) {
+                                break;
+                            }
+
+                            $innerValue .= $innerValues[$i]['next_glue_raw'];
+
+                            ++$i;
+                        }
+
+                        $this->innerTypeExpressions[] = [
+                            'start_index' => $innerStartIndex,
+                            'expression' => $this->inner($innerValue),
+                        ];
+                    }
+                }
+
+                if ('~' === $this->typesGlue) {
+                    $this->isGluedType = false; // subtraction is not commutative
+                }
 
                 return;
             }
         }
 
-        $this->isUnionType = false;
+        $this->isGluedType = false;
         $this->typesGlue = '|';
```

Comment:
```suggestion
```

Glue should be stored only if it's actually found in the type expression, it should remain `null` for non-composite types.

Reply:
reverted completely because of `~` not implemented in this PR - https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8210#discussion_r1765400782